File: src/front-end/src/infrastructure/transport/data-channel-framing.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { CHUNK_PAYLOAD_MAX, ChunkReassembler, packMessage } from './data-channel-framing';

function makePayload(size: number, fill = 0xab): Uint8Array {
  const payload = new Uint8Array(size);
  for (let i = 0; i < size; i++) {
    payload[i] = (i ^ fill) & 0xff;
  }
  return payload;
}

function roundTrip(msgId: number, payload: Uint8Array): Uint8Array | null {
  const reassembler = new ChunkReassembler();
  const chunks = packMessage(msgId, payload);
  let result: Uint8Array | null = null;
  for (const chunk of chunks) {
    result = reassembler.pushChunk(chunk);
  }
  return result;
}

// Verify byte-for-byte equality for small payloads; for large payloads only
// check size + boundary bytes to avoid enumerating millions of Uint8Array
// integer keys in vitest's deep-equality serializer (causes OOM at 16 MiB).
function assertRoundTrip(payload: Uint8Array): void {
  const result = roundTrip(1, payload);
  expect(result).not.toBeNull();
  if (payload.byteLength <= CHUNK_PAYLOAD_MAX + 1) {
    expect(result).toEqual(payload);
  } else {
    expect(result!.byteLength).toBe(payload.byteLength);
    expect(result![0]).toBe(payload[0]);
    expect(result![payload.byteLength - 1]).toBe(payload[payload.byteLength - 1]);
  }
}

describe('packMessage / ChunkReassembler round-trips', () => {
  it.each([
    ['1 byte', 1],
    ['CHUNK_PAYLOAD_MAX - 1', CHUNK_PAYLOAD_MAX - 1],
    ['CHUNK_PAYLOAD_MAX', CHUNK_PAYLOAD_MAX],
    ['CHUNK_PAYLOAD_MAX + 1', CHUNK_PAYLOAD_MAX + 1],
    ['1 MiB', 1 * 1024 * 1024],
    ['16 MiB', 16 * 1024 * 1024],
  ])('Success_roundTrip_%s', (_label, size) => {
    assertRoundTrip(makePayload(size));
  });

  it('Success_emptyPayload', () => {
    const payload = new Uint8Array(0);
    const result = roundTrip(1, payload);
    expect(result).not.toBeNull();
    expect(result!.byteLength).toBe(0);
  });
});

describe('ChunkReassembler chunk ordering', () => {
  it('Success_outOfOrderChunkArrival', () => {
    const payload = makePayload(CHUNK_PAYLOAD_MAX * 3);
    const chunks = packMessage(1, payload);
    expect(chunks).toHaveLength(3);

    const sut = new ChunkReassembler();
    expect(sut.pushChunk(chunks[2])).toBeNull();
    expect(sut.pushChunk(chunks[0])).toBeNull();
    const result = sut.pushChunk(chunks[1]);
    expect(result).not.toBeNull();
    expect(result).toEqual(payload);
  });

  it('Success_interleavedMessages', () => {
    const payload1 = makePayload(CHUNK_PAYLOAD_MAX * 2, 0x11);
    const payload2 = makePayload(CHUNK_PAYLOAD_MAX * 2, 0x22);
    const chunks1 = packMessage(1, payload1);
    const chunks2 = packMessage(2, payload2);

    const sut = new ChunkReassembler();
    expect(sut.pushChunk(chunks1[0])).toBeNull();
    expect(sut.pushChunk(chunks2[0])).toBeNull();
    const result1 = sut.pushChunk(chunks1[1]);
    const result2 = sut.pushChunk(chunks2[1]);

    expect(result1).not.toBeNull();
    expect(result2).not.toBeNull();
    expect(result1!.byteLength).toBe(payload1.byteLength);
    expect(result2!.byteLength).toBe(payload2.byteLength);
    expect(result1![0]).toBe(payload1[0]);
    expect(result2![0]).toBe(payload2[0]);
  });

  it('Edge_duplicateChunkIgnored', () => {
    const payload = makePayload(CHUNK_PAYLOAD_MAX * 2);
    const chunks = packMessage(1, payload);
    const sut = new ChunkReassembler();
    expect(sut.pushChunk(chunks[0])).toBeNull();
    expect(sut.pushChunk(chunks[0])).toBeNull(); // duplicate
    const result = sut.pushChunk(chunks[1]);
    expect(result).not.toBeNull();
    expect(result).toEqual(payload);
  });
});

describe('ChunkReassembler error paths', () => {
  it('Error_oversizedMessageThrows', () => {
    const payload = makePayload(16 * 1024 * 1024 + 1);
    expect(() => packMessage(1, payload)).toThrow();
  });

  it('Error_corruptedMagicDropsChunk', () => {
    const payload = makePayload(10);
    const chunks = packMessage(1, payload);
    const corrupted = chunks[0].slice(0);
    new Uint8Array(corrupted)[0] = 0x00;
    const sut = new ChunkReassembler();
    expect(sut.pushChunk(corrupted)).toBeNull();
  });

  it('Error_unknownVersionDropsChunk', () => {
    const payload = makePayload(10);
    const chunks = packMessage(1, payload);
    const corrupted = chunks[0].slice(0);
    new DataView(corrupted).setUint8(4, 99);
    const sut = new ChunkReassembler();
    expect(sut.pushChunk(corrupted)).toBeNull();
  });

  it('Error_bufferTooSmallDropsChunk', () => {
    const tiny = new ArrayBuffer(4);
    const sut = new ChunkReassembler();
    expect(sut.pushChunk(tiny)).toBeNull();
  });
});

describe('ChunkReassembler timeout', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('Error_timeoutDiscardsPartialMessage', () => {
    const payload = makePayload(CHUNK_PAYLOAD_MAX * 2);
    const chunks = packMessage(1, payload);
    const sut = new ChunkReassembler();

    expect(sut.pushChunk(chunks[0])).toBeNull();

    // Advance past the 5 s assembly timeout
    vi.advanceTimersByTime(6_000);

    // Second chunk arrives after timeout: the in-flight message was discarded,
    // so a new incomplete entry is created — still missing chunk[0] → null
    expect(sut.pushChunk(chunks[1])).toBeNull();
  });
});

File: src/cpp_accelerator/adapters/webrtc/data_channel_framing_test.cpp
#include "src/cpp_accelerator/adapters/webrtc/data_channel_framing.h"

#include <chrono>
#include <cstddef>
#include <thread>
#include <vector>

#include <gtest/gtest.h>

using namespace jrb::adapters::webrtc;

namespace {

std::vector<std::byte> MakePayload(size_t size,
                                    std::byte fill = std::byte{0xab}) {
  std::vector<std::byte> v(size);
  for (size_t i = 0; i < size; ++i) {
    v[i] = static_cast<std::byte>((i ^ static_cast<size_t>(fill)) & 0xffu);
  }
  return v;
}

std::optional<std::vector<std::byte>> RoundTrip(
    uint32_t msg_id, const std::vector<std::byte>& payload) {
  const auto chunks =
      PackMessage(msg_id, std::span<const std::byte>(payload));
  ChunkReassembler reassembler;
  std::optional<std::vector<std::byte>> result;
  for (const auto& chunk : chunks) {
    result = reassembler.PushChunk(std::span<const std::byte>(chunk));
  }
  return result;
}

}  // namespace

TEST(DataChannelFramingTest, Success_RoundTrip1Byte) {
  // Arrange
  const auto payload = MakePayload(1);
  // Act
  const auto result = RoundTrip(1, payload);
  // Assert
  ASSERT_TRUE(result.has_value());
  EXPECT_EQ(*result, payload);
}

TEST(DataChannelFramingTest, Success_RoundTripChunkPayloadMaxMinus1) {
  const auto payload = MakePayload(kChunkPayloadMax - 1);
  const auto result = RoundTrip(1, payload);
  ASSERT_TRUE(result.has_value());
  EXPECT_EQ(*result, payload);
}

TEST(DataChannelFramingTest, Success_RoundTripChunkPayloadMax) {
  const auto payload = MakePayload(kChunkPayloadMax);
  const auto result = RoundTrip(1, payload);
  ASSERT_TRUE(result.has_value());
  EXPECT_EQ(*result, payload);
}

TEST(DataChannelFramingTest, Success_RoundTripChunkPayloadMaxPlus1) {
  const auto payload = MakePayload(kChunkPayloadMax + 1);
  const auto result = RoundTrip(1, payload);
  ASSERT_TRUE(result.has_value());
  EXPECT_EQ(*result, payload);
}

TEST(DataChannelFramingTest, Success_RoundTrip1MiB) {
  const auto payload = MakePayload(1 * 1024 * 1024);
  const auto result = RoundTrip(1, payload);
  ASSERT_TRUE(result.has_value());
  EXPECT_EQ(*result, payload);
}

TEST(DataChannelFramingTest, Success_RoundTrip16MiB) {
  const auto payload = MakePayload(16 * 1024 * 1024);
  const auto result = RoundTrip(1, payload);
  ASSERT_TRUE(result.has_value());
  EXPECT_EQ(*result, payload);
}

TEST(DataChannelFramingTest, Success_OutOfOrderChunkArrival) {
  // Arrange
  const size_t size = static_cast<size_t>(kChunkPayloadMax) * 3;
  const auto payload = MakePayload(size);
  const auto chunks =
      PackMessage(1, std::span<const std::byte>(payload));
  ASSERT_EQ(chunks.size(), 3u);

  // Act — deliver chunks out of order: 2, 0, 1
  ChunkReassembler sut;
  EXPECT_FALSE(sut.PushChunk(chunks[2]).has_value());
  EXPECT_FALSE(sut.PushChunk(chunks[0]).has_value());
  const auto result = sut.PushChunk(chunks[1]);

  // Assert
  ASSERT_TRUE(result.has_value());
  EXPECT_EQ(*result, payload);
}

TEST(DataChannelFramingTest, Success_InterleavedMessages) {
  // Arrange
  const auto payload1 = MakePayload(kChunkPayloadMax * 2, std::byte{0x11});
  const auto payload2 = MakePayload(kChunkPayloadMax * 2, std::byte{0x22});
  const auto chunks1 =
      PackMessage(1, std::span<const std::byte>(payload1));
  const auto chunks2 =
      PackMessage(2, std::span<const std::byte>(payload2));
  ASSERT_EQ(chunks1.size(), 2u);
  ASSERT_EQ(chunks2.size(), 2u);

  // Act — interleave: msg1[0], msg2[0], msg1[1], msg2[1]
  ChunkReassembler sut;
  EXPECT_FALSE(sut.PushChunk(chunks1[0]).has_value());
  EXPECT_FALSE(sut.PushChunk(chunks2[0]).has_value());
  const auto result1 = sut.PushChunk(chunks1[1]);
  const auto result2 = sut.PushChunk(chunks2[1]);

  // Assert
  ASSERT_TRUE(result1.has_value());
  ASSERT_TRUE(result2.has_value());
  EXPECT_EQ(*result1, payload1);
  EXPECT_EQ(*result2, payload2);
}

TEST(DataChannelFramingTest, Error_PayloadTooLargeThrows) {
  const auto payload = MakePayload(kMaxMessageBytes + 1);
  EXPECT_THROW(PackMessage(1, std::span<const std::byte>(payload)),
               std::invalid_argument);
}

TEST(DataChannelFramingTest, Error_CorruptedMagicDropsChunk) {
  // Arrange
  const auto payload = MakePayload(10);
  auto chunks = PackMessage(1, std::span<const std::byte>(payload));
  ASSERT_FALSE(chunks.empty());
  chunks[0][0] = std::byte{0x00};  // corrupt first magic byte

  // Act & Assert
  ChunkReassembler sut;
  EXPECT_FALSE(sut.PushChunk(std::span<const std::byte>(chunks[0])).has_value());
}

TEST(DataChannelFramingTest, Error_UnknownVersionDropsChunk) {
  // Arrange
  const auto payload = MakePayload(10);
  auto chunks = PackMessage(1, std::span<const std::byte>(payload));
  ASSERT_FALSE(chunks.empty());
  chunks[0][4] = std::byte{99};  // unknown version byte

  // Act & Assert
  ChunkReassembler sut;
  EXPECT_FALSE(sut.PushChunk(std::span<const std::byte>(chunks[0])).has_value());
}

TEST(DataChannelFramingTest, Error_BufferTooSmallDropsChunk) {
  const std::vector<std::byte> tiny(4, std::byte{0x00});
  ChunkReassembler sut;
  EXPECT_FALSE(sut.PushChunk(std::span<const std::byte>(tiny)).has_value());
}

TEST(DataChannelFramingTest, Error_TimeoutDiscardsPartialMessage) {
  // Arrange — use a 1 ms timeout so we don't have to sleep 5 seconds
  const size_t size = static_cast<size_t>(kChunkPayloadMax) * 2;
  const auto payload = MakePayload(size);
  const auto chunks =
      PackMessage(1, std::span<const std::byte>(payload));
  ASSERT_EQ(chunks.size(), 2u);

  ChunkReassembler sut{std::chrono::milliseconds{1}};

  // Act — push first chunk, let timeout expire, push second
  EXPECT_FALSE(sut.PushChunk(chunks[0]).has_value());
  std::this_thread::sleep_for(std::chrono::milliseconds{5});
  // The second chunk arrives after timeout. The evicted message is gone, so
  // a new entry is created with only chunk[1] present — still incomplete.
  EXPECT_FALSE(sut.PushChunk(chunks[1]).has_value());
}

TEST(DataChannelFramingTest, Edge_DuplicateChunkIgnored) {
  // Arrange
  const auto payload = MakePayload(kChunkPayloadMax * 2);
  const auto chunks =
      PackMessage(1, std::span<const std::byte>(payload));
  ASSERT_EQ(chunks.size(), 2u);

  // Act — send chunk[0] twice then chunk[1]
  ChunkReassembler sut;
  EXPECT_FALSE(sut.PushChunk(chunks[0]).has_value());
  EXPECT_FALSE(sut.PushChunk(chunks[0]).has_value());  // duplicate
  const auto result = sut.PushChunk(chunks[1]);

  // Assert
  ASSERT_TRUE(result.has_value());
  EXPECT_EQ(*result, payload);
}
